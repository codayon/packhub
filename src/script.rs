use anyhow::anyhow;
use askama::Template;
use axum::{Router, extract::Path, routing::get};

use crate::{error::AppError, state::AppState};

#[derive(Template)]
#[template(path = "apt-script.sh", escape = "none")]
struct AptScript<'a> {
    host: &'a str,
    distro: &'a str,
    owner: &'a str,
    repo: &'a str,
}

fn generate_apt_script(distro: &str, owner: &str, repo: &str) -> String {
    let host = dotenvy::var("PACKHUB_DOMAIN").unwrap();
    let script = AptScript {
        host: &host,
        distro,
        owner,
        repo,
    };
    script.render().unwrap()
}

#[derive(Template)]
#[template(path = "rpm-script.sh", escape = "none")]
struct RPMScript<'a> {
    host: &'a str,
    owner: &'a str,
    repo: &'a str,
}

fn generate_rpm_script(owner: &str, repo: &str) -> String {
    let host = dotenvy::var("PACKHUB_DOMAIN").unwrap();
    let script = RPMScript {
        host: &host,
        owner,
        repo,
    };
    script.render().unwrap()
}

async fn script_handler(
    Path((distro, owner, repo)): Path<(String, String, String)>,
) -> Result<String, AppError> {
    match distro.as_str() {
        "ubuntu" | "debian" => Ok(generate_apt_script(&distro, &owner, &repo)),
        "rpm" => Ok(generate_rpm_script(&owner, &repo)),
        _ => Err(anyhow!("Script Generation: Unsupported distro: {}", distro).into()),
    }
}

pub fn script_routes() -> Router<AppState> {
    Router::new().route("/{distro}/github/{owner}/{repo}", get(script_handler))
}

#[cfg(test)]
mod tests {
    use insta::assert_snapshot;

    use super::*;

    #[test]
    fn test_script_generation_apt() {
        let apt_script = generate_apt_script("ubuntu", "OpenBangla", "OpenBangla-Keyboard");
        assert_snapshot!(apt_script);
    }

    #[test]
    fn test_script_generation_rpm() {
        let script = generate_rpm_script("OpenBangla", "OpenBangla-Keyboard");
        assert_snapshot!(script);
    }
}
