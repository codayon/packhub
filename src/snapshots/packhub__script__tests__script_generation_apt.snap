---
source: src/script.rs
expression: apt_script
---
#!/bin/sh

echo "Welcome to package key and repository setup script for OpenBangla-Keyboard"
echo "This script will add the repository key and repository to your system."
echo "Please make sure you have sudo access to run this script."
echo
echo "Installing GnuPG for key verification..."
sudo apt-get install gpg -y
echo
echo "Downloading and installing the repository key..."
wget -qO- http://localhost:3000/keys/packhub.asc | gpg --dearmor > packhub.gpg
sudo install -D -o root -g root -m 644 packhub.gpg /etc/apt/keyrings/packhub.gpg
echo
echo "Adding the repository to your system..."
echo "deb [signed-by=/etc/apt/keyrings/packhub.gpg] http://localhost:3000/apt/github/OpenBangla/OpenBangla-Keyboard stable main" > /etc/apt/sources.list.d/OpenBangla-Keyboard.list
echo 
echo "Updating package lists..."
sudo apt-get update
